tools/nimgrep.nim

```diff
@@ -1298,28 +1361,42 @@ checkOptions({optCount, optReplace}, "count", "replace")
 checkOptions({optPeg, optRegex}, "peg", "re")
 checkOptions({optIgnoreCase, optIgnoreStyle}, "ignore_case", "ignore_style")
 checkOptions({optFilenames, optReplace}, "filenames", "replace")
+checkOptions({optPipe, optStdin}, "-", "stdin")
+checkOptions({optPipe, optFilenames}, "-", "filenames")
+checkOptions({optPipe, optConfirm}, "-", "confirm")
+checkOptions({optPipe, optRecursive}, "-", "recursive")
 
 linesBefore = max(linesBefore, linesContext)
 linesAfter  = max(linesAfter,  linesContext)
 
+if optPipe in options and paths.len != 0:
+  reportError("both - and paths are specified")
+
 if optStdin in options:
   searchOpt.pattern = ask("pattern [ENTER to exit]: ")
   if searchOpt.pattern.len == 0: quit(0)
   if optReplace in options:
     replacement = ask("replacement [supports $1, $# notations]: ")
 
-if searchOpt.pattern.len == 0:
+if optReplace in options and not replacementSet:
+  reportError("provide REPLACEMENT as second argument (use \"\" for empty one)")
+if optReplace in options and paths.len == 0 and optPipe notin options:
+  reportError("provide paths for replacement explicitly (use . for current directory)")
+
+if searchOpt.pattern == "" and optFilenames notin options:
   reportError("empty pattern was given")
 else:
-  if paths.len == 0:
+  if paths.len == 0 and optPipe notin options:
     paths.add(".")
-  if nWorkers == 0:
+  if optPipe in options or nWorkers == 0:
     run1Thread()
   else:
     runMultiThread()
   if gVar.errors != 0:
     printError $gVar.errors & " errors"
-  printBold($gVar.matches & " matches")
-  stdout.write("\n")
+  if searchOpt.pattern != "":
+    # PATTERN allowed to be empty if --filenames is given
+    printBold($gVar.matches & " matches")
+    stdout.write("\n")
   if gVar.errors != 0:
     quit(1)
```

Comment:
unrelated TODO for future PR: `reportError`?

Reply:
it reports error 4 lines above. It's done this way for matches to always be last:
```
<M> errors 
<N> matches
```